engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/SortNode.java

```diff
@@ -0,0 +1,102 @@
+  private final ConditionProfile consProfile = ConditionProfile.createCountingProfile();
+  private final ConditionProfile lessProfile = ConditionProfile.createCountingProfile();
+  private final ConditionProfile equalsProfile = ConditionProfile.createCountingProfile();
+  private final ConditionProfile greaterProfile = ConditionProfile.createCountingProfile();
+
+  abstract Object execute(VirtualFrame frame, Object _this, Object comparator);
+
+  static SortNode build() {
+    return SortNodeGen.create();
+  }
+
+  @Specialization
+  Object doSort(
+      VirtualFrame frame,
+      Array _this,
+      Object comparator,
+      @Cached("buildInvokeNode()") InvokeCallableNode invokeNode,
+      @CachedContext(Language.class) ContextReference<Context> ctxRef,
+      @Cached("ctxRef.get().getBuiltins().ordering().newLess()") Atom less,
+      @Cached("ctxRef.get().getBuiltins().ordering().newEqual()") Atom equal,
+      @Cached("ctxRef.get().getBuiltins().ordering().newGreater()") Atom greater) {
+    Comparator<Object> compare =
+        (l, r) -> {
+          Stateful result =
+              invokeNode.execute(comparator, frame, EmptyMap.create(), new Object[] {l, r});
+          if (TypesGen.isAtom(result.getValue())) {
+            Atom atom = TypesGen.asAtom(result.getValue());
+            if (lessProfile.profile(atom == less)) {
+              return -1;
+            } else if (equalsProfile.profile(atom == equal)) {
+              return 0;
+            } else if (greaterProfile.profile(atom == greater)) {
+              return 1;
+            }
+          }
+
+          CompilerDirectives.transferToInterpreter();
+          var ordering = ctxRef.get().getBuiltins().ordering().ordering();
+          throw new PanicException(
+              ctxRef.get().getBuiltins().error().makeTypeError(ordering, result.getValue()), this);
+        };
+    doSort(_this.getItems(), compare);
+    return ctxRef.get().getBuiltins().nothing().newInstance();
+  }
+
+  @Specialization
+  Object doAtomThis(
+      VirtualFrame frame,
+      Atom _this,
+      Object that,
+      @CachedContext(Language.class) ContextReference<Context> ctxRef,
+      @Cached("ctxRef.get().getBuiltins().mutable().array()") AtomConstructor array) {
+    if (consProfile.profile(array == _this.getConstructor())) {
+      return ctxRef.get().getBuiltins().nothing().newInstance();
+    } else {
+      CompilerDirectives.transferToInterpreter();
+      throw new PanicException(
+          ctxRef.get().getBuiltins().error().makeTypeError(array, _this), this);
+    }
+  }
+
+  @TruffleBoundary
+  void doSort(Object[] items, Comparator<Object> compare) {
+    Arrays.sort(items, compare);
+  }
+
+  InvokeCallableNode buildInvokeNode() {
+    CallArgumentInfo[] callArguments = {new CallArgumentInfo(), new CallArgumentInfo()};
+    return InvokeCallableNode.build(
+        callArguments, DefaultsExecutionMode.IGNORE, ArgumentsExecutionMode.EXECUTE);
```

Comment:
```suggestion
        callArguments, DefaultsExecutionMode.EXECUTE, ArgumentsExecutionMode.PRE_EXECUTED);
```

Reply:
1. We want defaulted arguments to work normally for this comparator.
2. We always call it with fully evaluated objects.